### PR TITLE
WebGLRenderer: Add "readRenderTargetPixelsAsync" function

### LINKED ISSUE #22779: WebGL2: Asynchronous Readback with PIXEL_PACK_BUFFERs
**Is your feature request related to a problem? Please describe.**

WebGL2 (which is now supported by all modern browsers) offers powerful General Purpose GPU capabilities.  three.js already implements an impressive array of these features (like `WebGLMultipleRenderTargets`).   This GPGPU suite is nearly complete, but it's missing a performant way of reading texture data back from the GPU.  `renderer.readRenderTargetPixels()` waits for a sync with the GPU before the download starts, costing tens of milliseconds per-call on PC and more on mobile.

**Describe the solution you'd like**

WebGL2 adds the ability to bind `_gl.PIXEL_PACK_BUFFER`s, which allow for asynchronous readback from the GPU Texture Buffers, largely eliminating the performance impact of GPU -> Code.     This should open up all kinds of GPU Physics (and other GPGPU) Solutions.

**Describe alternatives you've considered**

[Luma.gl implemented async readback a few years ago](https://github.com/visgl/luma.gl/pull/450)
[Babylon.js also has some functionality for this](https://forum.babylonjs.com/t/speeding-up-readpixels/12739/)
([One implementation in Babylon...](https://forum.babylonjs.com/t/read-pixels-from-texture-asynchronously/19371/9))
[Unity has a similar mechanism for this](https://docs.unity3d.com/ScriptReference/Rendering.AsyncGPUReadback.html)

**Additional context**

My immediate use-case would be to speed up this [GPU Finite Element Modelling simulation](https://zalo.github.io/TetSim/).
If I can get everything running fast a stable on all platforms, I'd also like to contribute a new GPUComputationRenderer (which would support `WebGLMultipleRenderTargets` and separate the concept of "passes" from "variables", allowing for multiple passes over the same data).

### PR DESCRIPTION
**Pull request by gkjohnson**
Fix #22779
Based on #24466

**Description**

Adds a `readRenderTargetPixelsAsync` to WebGLRenderer which supports asynchronously reading pixels from a render target by waiting until the read commands have finished before providing the array buffer. The `webgl_interactive_cubes_gpu` example has also been updated to demonstrate the functions use.

Here's a quick look at the stall times when reading data in the interactive cubes example:

<details>
  <summary>Benchmarks</summary>

### Reading 1*1 Pixel Texture
**readRenderTargetPixels**

~2.2 - 2.5 ms stall when calling `readRenderTargetPixels( ... )`

**readRenderTargetPixelsAsync**

~0.5 - 1.0 ms stall when initially calling `readRenderTargetPixelsAsync( ... )`
~8.8 - 10.0 ms until the data is ready (asynchronous) or 1 or 2 frames on my 120 fps display.

### Reading 1024*1024 Pixel Texture
**readRenderTargetPixels**

~10.0 - 11.0 ms stall when calling `readRenderTargetPixels( ... )`

**readRenderTargetPixelsAsync**

~0.5 - 1.2 ms stall when initially calling `readRenderTargetPixelsAsync( ... )`
~12.0 - 13.0 ms until the data is ready (asynchronous) or 2 frames on my 120 fps display.

</details>

cc @sciecode I've updated and simplified some of the code from #24466 and removed some of the tangential functions and options. If they're needed perhaps we can add them in future PRs but the use wasn't immediately clear to me. Unfortunately I couldn't include your contributed commits in this PR because the branch had been deleted.

_This contribution is funded by the [Cesium GIS Ecosystem Grant](https://cesium.com/cesium-ecosystem-grants/grant-directory/)_

### CODE REVIEW

**Comment by sciecode** [comment]
LGTM - Removed functions were used to test alternative APIs, they aren't required.

nit: I found empirically that a fixed 8ms is bit too large of an interval for some use cases, I have used 4ms or even 2ms in some cases. Since the probe function isn't too expensive, having a higher refresh rate can decrease waiting time quite significantly. Ideally this would be a user defined parameter, but I think a 4ms default is a nice compromise for now.

Other than that, should be good to go.

File: src/renderers/WebGLRenderer.js [hunk 1]
```diff
@@ -2357,6 +2357,85 @@ class WebGLRenderer {
 
 		};
 
+		this.readRenderTargetPixelsAsync = async function ( renderTarget, x, y, width, height, buffer, activeCubeFaceIndex ) {
```
Comment (Mugen87) [inline src/renderers/WebGLRenderer.js:2360]: Apart from documenting the new method the PR looks good!

**Comment by gkjohnson** [comment]
Thanks @sciecode I've adjusted the value to 4ms

> Ideally this would be a user defined parameter

If we see issues or needs in the future maybe we can look at add some more options.

File: src/renderers/WebGLRenderer.js (hunk 1 above)
Comment (gkjohnson) [reply src/renderers/WebGLRenderer.js:2360]: Added!

**Review by Mugen87 — APPROVED** [review]

**Pull request merged**